Add FLIP for improving Cadence numeric supertypes (#729)
onflow/flow: flips/20211210-cadence-numeric-supertypes.md

# Improve Cadence Numeric Supertypes

| Status        | Proposed       |
:-------------- |:---------------------------------------------------- |
| **FLIP #**    | N/A|
| **Author(s)** | Supun Setunga (supun.setunga@dapperlabs.com)       |
| **Sponsor**   | Supun Setunga (supun.setunga@dapperlabs.com)       |
| **Updated**   | 2021-12-13                                           |

## Objective

Statically disallow arithmetic operations on numeric supertypes, as they may or may not succeed at
run-time.
The idea is to make the current behavior that happens under the hood, more explicit to the developers,
and by doing so, help them to avoid any unintentional/unforeseen behaviors.

## Motivation

Cadence type system consists of several numeric supertypes, such as:
- Number
- SignedNumber
- FixedPoint
- SignedFixedPoint
- Integer
- SignedInteger

Currently, the arithmetic operations are also allowed on those types (e.g. `+`, `-` `*`, `/`, etc.).

However, it is not always guaranteed that these operations will succeed at run-time.

> e.g:
> If the two numeric values are of same type, then the operation will succeed.

Reviewer: ```suggestion
For example, if the two numeric values are of same type, then the operation will succeed:
```

```cadence
let x: Integer = 3 as Int8
let y: Integer = 4 as Int8
let z: Integer = x + y
```

> If the two numeric values are of different types, then there will be a run-time panic.
> ```cadence

Reviewer: ```suggestion

However, if the two numeric values are of different types, then there will be a run-time panic.
```cadence
```

let x: Integer = 3 as Int8
let y: Integer = 4 as UInt16

// Panics at run-time, since the two numbers are of different types
let z: Integer = x + y
```

This leads to situations where developers would end up writing codes that were assumed to work,
but instead, fails at run-time for certain combination of inputs.
It is not intuitive enough for developers that their codes may not work for all inputs.

## User Benefit
It would be more intuitive and explicit to the developers and to anyone who read the code, about
what's happening during arithmetic operations on numeric supertypes. 
Helps them to avoid any unintentional/unforeseen behaviors.

## Design Proposal

The proposed solution is to statically disallow arithmetic operations on numeric supertypes.

> ```
> let x: Integer = 3 as Int8
> let y: Integer = 4 as Int8
> 
> let z: Integer = x + y     // Semantic error
> ```

Reviewer: ```suggestion

```cadence
let x: Integer = 3 as Int8
let y: Integer = 4 as Int8

let z: Integer = x + y     // Static error
```
```

Reviewer 2: Shouldn't one of those be a different type? Why would two `Int8` numbers being added cause a static error?

Developers will have to explicitly force-cast it to the desired type before they do the arithmetic
operation.

> ```

Reviewer: ```suggestion

```cadence
```

let x: Integer = 3 as Int8

> let y: Integer = 4 as Int8

Reviewer 2: Should one of these be a different type?

Author: This demonstrates the new behavior once the change is made. This example is for the successful scenario, where runtime values are of the same type.

To clarify more: Here even though the two values `3` and `4` are of `Int8` type, the two variables `x` and `y` are of `Integer` supertype. Hence the `x + y` is not allowed. Instead, to make it work, users will have to cast it before add-operation (`(x as! Int8) + (y as! Int8)`). Then it would work as before.
